storage/replica.go

```diff
@@ -511,8 +536,22 @@ func (r *Replica) setDesc(desc *roachpb.RangeDescriptor) error {
 
 // setDescWithoutProcessUpdate updates the range descriptor without calling
 // processRangeDescriptorUpdate.
-func (r *Replica) setDescWithoutProcessUpdate(desc *roachpb.RangeDescriptor) {
-	atomic.StorePointer(&r.desc, unsafe.Pointer(desc))
+func (r *Replica) setDescWithoutProcessUpdate(desc *roachpb.RangeDescriptor) error {
+	r.Lock()
+	defer r.Unlock()
+	return r.setDescWithoutProcessUpdateLocked(desc)
+}
+
+// setDescWithoutProcessUpdateLocked updates the range descriptor without
+// calling processRangeDescriptorUpdate.
+// setDescWithoutProcessUpdateLocked requires that the replica lock is held.
+func (r *Replica) setDescWithoutProcessUpdateLocked(desc *roachpb.RangeDescriptor) error {
+	if desc.RangeID != r.RangeID {
+		return util.Errorf("range descriptor ID (%d) does not match replica's range ID (%d)",
```

Comment:
I'd make this a panic; it shouldn't be possible to ever overwrite a range id like this.